internal/datastore/crdb/pool/balancer.go

```diff
@@ -0,0 +1,214 @@
+	IsClosed() bool
+}
+
+// balanceablePool is an interface that a pool must implement to allow its
+// connections to be balanced by the balancer.
+type balanceablePool[P balancePoolConn[C], C balanceConn] interface {
+	ID() string
+	AcquireAllIdle(ctx context.Context) []P
+	Node(conn C) uint32
+	GC(conn C) uint32
+	MaxConns() uint32
+	Range(func(conn C, nodeID uint32))
+}
+
+// NodeConnectionBalancer attempts to keep the connections managed by a RetryPool balanced between healthy nodes in
+// a Cockroach cluster.
+// It asynchronously processes idle connections, and kills any to nodes that have too many. When the pool reconnects,
+// it will have a different balance of connections, and over time the balancer will bring the counts close to equal.
+type NodeConnectionBalancer struct {
+	nodeConnectionBalancer[*pgxpool.Conn, *pgx.Conn]
+}
+
+// NewNodeConnectionBalancer builds a new nodeConnectionBalancer for a given connection pool and health tracker.
+func NewNodeConnectionBalancer(pool *RetryPool, healthTracker *NodeHealthTracker, interval time.Duration) *NodeConnectionBalancer {
+	return &NodeConnectionBalancer{*newNodeConnectionBalancer[*pgxpool.Conn, *pgx.Conn](pool, healthTracker, interval)}
+}
+
+// nodeConnectionBalancer is generic over underlying connection types for
+// testing purposes. Callers should use the exported NodeConnectionBalancer
+type nodeConnectionBalancer[P balancePoolConn[C], C balanceConn] struct {
+	ticker        *time.Ticker
+	pool          balanceablePool[P, C]
+	healthTracker *NodeHealthTracker
+	rnd           *rand.Rand
+	seed          int64
+}
+
+// newNodeConnectionBalancer is generic over underlying connection types for
+// testing purposes. Callers should use the exported NewNodeConnectionBalancer.
+func newNodeConnectionBalancer[P balancePoolConn[C], C balanceConn](pool balanceablePool[P, C], healthTracker *NodeHealthTracker, interval time.Duration) *nodeConnectionBalancer[P, C] {
+	seed := int64(new(maphash.Hash).Sum64())
+	return &nodeConnectionBalancer[P, C]{
+		ticker:        time.NewTicker(interval),
+		healthTracker: healthTracker,
+		pool:          pool,
+		seed:          seed,
+		rnd:           rand.New(rand.NewSource(seed)),
+	}
+}
+
+// Prune starts periodically checking idle connections and killing ones that are determined to be unbalanced.
+func (p *nodeConnectionBalancer[P, C]) Prune(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			p.ticker.Stop()
+			return
+		case <-p.ticker.C:
+			p.pruneConnections(ctx)
+			runtime.GC()
+		}
+	}
+}
+
+// pruneConnections prunes connections to nodes that have more than MaxConns/(# of nodes)
+// This causes the pool to reconnect, which over time will lead to a balanced number of connections
+// across each node.
+func (p *nodeConnectionBalancer[P, C]) pruneConnections(ctx context.Context) {
+	conns := p.pool.AcquireAllIdle(ctx)
+	defer func() {
+		// release all acquired idle conns back
+		for _, c := range conns {
+			c.Release()
+		}
+	}()
```

Comment:
I actually think this calculation is very fast - there are mutexes, but none of them are ever held for long, and there is a sort, but only of a small set of uints.

I could add a metric and we could monitor if this method is adding too much contention?

Reply:
we discussed this out of band, some considerations:
- hot-spot caching means there are spikes of pool usage that could line up with the balancer locking all idle conns
- in high contention scenarios this wouldn't be a problem since the balancer wouldn't get many idle connections
- we'll add the metric to evaluate the duration of the method and determine if this is a problem at all

Reply:
added the metric and ran some scenarios locally. A typical value is ~100μs, though I did see one spike to 10ms during an upgrade (but not consistently).